Development strategy roadmap

ajgpitch opened the pull request:
Document to outline strategy and ideas for development of qutip and associated packages.
In particular plans for the next major release.

ajgpitch commented:
The section on "what is QuTiP" might seem a bit random. It is a admission of defeat in my attempt to think of some way to make the qutip / QuTiP naming less confusing.

ajgpitch commented:
Note that this document includes ideas that have been fully discussed by the Admin team, but also proposals that have recently been dreamed up by myself alone. So much needs to be considered and agreed upon before any roadmap can be adopted. Hopefully this will provide a useful starting point for discussion and an opportunity to gain input from the Community.

nwlambert commented:
Nice summary! Some initial thoughts:

- To verbally distinguish QuTiP and qutip I propose we make the P in QuTiP silent.  :P 
- I like the idea of bringing in and maintaining the sympsi stuff, might be a nice gsoc project?  I think Eunjong did work on this too, so might be worth talking to him (https://github.com/eunjongkim) as well as RJ.  
- Worth listing HEOM as a goal for v5? 

Something on my mind for a while that might make a straightforward gsoc project/milestone are some updates to the bloch-redfield solver with a non-secular variant. I have a rough python version of this already locally, but would like to generalize it/cythonise it and make it officially part of brsolve.

ajgpitch commented:
@nwlambert 
Eunjong is definitely listed as a contributor on the Sympsi site. I suppose I missed Eunjong out as I don't know him. I will fix that now.
We should definitely include the new HEOM as a workpackage. We should add it in for v.5 release plan. I suppose there is no harm in adding this in now, contrary to what I said above. You could either make a suggestion via GH Review or just send me some text and I'll do it.
We can do similar for the the BR plan too.

BoxiLi commented on dev_roadmap.md:
This is definitely a useful feature. But on the other hand, it might need quite some work. Seems that each quantum computing platform has there own developers on this. Because we only have a few developers, I'm wondering if we should compete on this?

ajgpitch replied:
Good point. I don't think it's high priority

BoxiLi commented on dev_roadmap.md:
```diff
@@ -0,0 +1,329 @@
+The QIP subpackage has been deemed out of scope (feature-wise). It also depends on `qutip.control` and hence would be out of scope for dependency reasons.
+A separate repository has already been made for qutip-qip.
+ 
+ #### qutip-symbolic
```
This is interesting. Actually, I'm doing some symbolic computing right now. Maybe I can help in fixing some bugs.

ajgpitch replied:
Cool. There is quite a lot of potential for symbolic stuff inside of qutip.qutip I think. I have been discussing with @hodgestar too, maybe you two should meet to discuss more

hodgestar replied:
@BoxiLi I would love to have a quick catch up about what you're working on and see if we can combine forces or work nicely in parallel.

BoxiLi reviewed (commented):
Many thanks for drafting the roadmap! These are just some comments or to discuss, not necessarily suggestions.

ajgpitch commented on dev_roadmap.md:
```diff
@@ -0,0 +1,329 @@
+ - QuTiP family packages. The main qutip package already has sub-packages, which are maintained in the main qutip repo. Any packages maitained by the QuTiP organisation will be called QuTiP 'family' packages. Sub-packages within qutip main will be called 'integrated' sub-packages. Some packages will be maintained in their own repos and installed separately within the main qutip folder structure t ...
+ - QuTiP affilliated packages. Other packages have been developed by others outside of the QuTiP organisation that work with, and are complementary to, qutip. The plan is to give some recognition to those that we deserve worthy of such [this needs clarification]. These packages will not be maintained by the QuTiP Team.
+ 
+ &#x1F537; Do we really need optional subpackages? It seems that are a bit fiddly and as we are side-stepping bw compat with a major version, we could just make QIP a separate package.
```
> That's a fair point. The intention of making it an optional package (plugin) was indeed for backward compatibility, to make it as close as possible to the old version. The same holds for `control` and `lattice` I guess? We can surely discuss this. But if we decide to be bold and make all of them separate package, we should let Sidhant know as soon as possible since this will reduce his work.

Yes, we should discuss this very soon and let @sarsid know what we decide.

ajgpitch commented on dev_roadmap.md:
```diff
@@ -0,0 +1,329 @@
+ 
+ --------------------------------------------------------------
+
+ ## Workpackages
```
Good idea. I will change this.

ajgpitch commented on dev_roadmap.md:
```diff
@@ -0,0 +1,329 @@
+ 
+ *main dev*: TBA
+ 
+ The qutip user documentation build files are to be moved to the qutip/qutip repo.
```
In their own repos

ajgpitch commented:
I have added a section on the HEOM project, and added it to the v.5 must haves. @nwlambert  please check you are happy with the wording. Let me know if there is something you wanted to add.

jakelishman commented on dev_roadmap.md:
```diff
@@ -0,0 +1,341 @@
+
+ ## Development Projects
+ 
+ ### data layer abstraction 
+ &#x1F535; tag: dl-abs
+ 
+ *status*: development [pretty much] complete
```
"pretty much" is a _little_ optimistic - there's still bits to do here in providing full control over how the data-layer dispatchers choose the most appropriate output type - but definitely the heaviest lifting is done.  I'm going to try and write the final design spec for the little remaining stuff this weekend.

ajgpitch replied:
Thanks @jakelishman 
I will update to better reflect the actual status

jakelishman commented:
I also have the overhaul of binary wheel distribution mostly ready, and maybe in a week or so's time I could do with arranging a meeting with someone to arrange for the PyPI secrets to be put in the right places.  I've been MIA from any development for a couple of months, and trying to get back on top of what I've got again (hence my little flurry of activity elsewhere).

jakelishman commented:
Oh, also we have to re-instate OpenMP on `dev.major`.  @Ericgig this is something we need to work out how we're going to do properly, especially since MKL and some other BLAS/LAPACK implementations do their own thing a little bit with regards to threading.

Cython gives us some nice-ish tools to handle OpenMP, and with us moving to package into binary wheels we ought to be able to have OpenMP compiled into all releases, but only active if a suitable OpenMP runtime library is available.  I _think_ OpenMP runtimes are generally cross-compiler compatible.

ajgpitch commented:
Thank you @jakelishman  for your comments. I will add another Project for the OpenMP support

ajgpitch commented:
I am merging this now. Further additions and edits can be made through future PR.
